openhcl/hcl/src/ioctl.rs

```diff
@@ -1521,57 +1618,160 @@ impl<T> Drop for ProcessorRunner<'_, T> {
+                | HvX64RegisterName::MsrMtrrPhysBaseC
+                | HvX64RegisterName::MsrMtrrPhysBaseD
+                | HvX64RegisterName::MsrMtrrPhysBaseE
+                | HvX64RegisterName::MsrMtrrPhysBaseF
+                | HvX64RegisterName::MsrMtrrPhysMask0
+                | HvX64RegisterName::MsrMtrrPhysMask1
+                | HvX64RegisterName::MsrMtrrPhysMask2
+                | HvX64RegisterName::MsrMtrrPhysMask3
+                | HvX64RegisterName::MsrMtrrPhysMask4
+                | HvX64RegisterName::MsrMtrrPhysMask5
+                | HvX64RegisterName::MsrMtrrPhysMask6
+                | HvX64RegisterName::MsrMtrrPhysMask7
+                | HvX64RegisterName::MsrMtrrPhysMask8
+                | HvX64RegisterName::MsrMtrrPhysMask9
+                | HvX64RegisterName::MsrMtrrPhysMaskA
+                | HvX64RegisterName::MsrMtrrPhysMaskB
+                | HvX64RegisterName::MsrMtrrPhysMaskC
+                | HvX64RegisterName::MsrMtrrPhysMaskD
+                | HvX64RegisterName::MsrMtrrPhysMaskE
+                | HvX64RegisterName::MsrMtrrPhysMaskF
+                | HvX64RegisterName::MsrMtrrFix64k00000
+                | HvX64RegisterName::MsrMtrrFix16k80000
+                | HvX64RegisterName::MsrMtrrFix16kA0000
+                | HvX64RegisterName::MsrMtrrFix4kC0000
+                | HvX64RegisterName::MsrMtrrFix4kC8000
+                | HvX64RegisterName::MsrMtrrFix4kD0000
+                | HvX64RegisterName::MsrMtrrFix4kD8000
+                | HvX64RegisterName::MsrMtrrFix4kE0000
+                | HvX64RegisterName::MsrMtrrFix4kE8000
+                | HvX64RegisterName::MsrMtrrFix4kF0000
+                | HvX64RegisterName::MsrMtrrFix4kF8000
+                | HvX64RegisterName::Dr0
+                | HvX64RegisterName::Dr1
+                | HvX64RegisterName::Dr2
+                | HvX64RegisterName::Dr3
+                | HvX64RegisterName::Dr6
+        )
+    }
+
+    #[cfg(guest_arch = "aarch64")]
+    fn is_kernel_managed(&self, _name: HvArm64RegisterName) -> bool {
+        false
+    }
+
+    fn set_reg(&mut self, regs: &[HvRegisterAssoc], vtl: GuestVtl) -> Result<(), Error> {
         if regs.is_empty() {
             return Ok(());
         }
-        // TODO GUEST_VSM
-        let vtl = Vtl::Vtl0;
+
         if let Some(sidecar) = &mut self.sidecar {
             sidecar
                 .set_vp_registers(vtl.into(), regs)
                 .map_err(Error::Sidecar)?;
         } else {
-            for regs in regs.chunks(MSHV_VP_MAX_REGISTERS) {
-                let hv_vp_register_args = mshv_vp_registers {
-                    count: regs.len() as i32,
-                    regs: regs.as_ptr().cast_mut(),
-                };
-                // SAFETY: IOCTL call with correct types.
-                unsafe {
-                    hcl_set_vp_register(self.hcl.mshv_vtl.file.as_raw_fd(), &hv_vp_register_args)
+            // TODO: group up to MSHV_VP_MAX_REGISTERS regs. The kernel
```

Comment:
We might as well go ahead and batch when none of the registers are kernel managed.

Reply:
My latest commits have a change to implement this. There's a bug that I haven't figured out: some WHP tests fail with `invalid vm state err=vp state is invalid`. This reliably goes away when I take out the register-batching change. Looking into it a bit more, right before that, something (VTL2, I think) is trying to inject a segment-not-present exception into PendingEvent0.

Reply:
I think this is a low priority for now, at least for production scenarios. Booting a Windows guest with Underhill (VBS-isolated with Guest VSM enabled), I see three calls to `set_reg`. There are many calls to `set_vp_registers`, but everything seems to be handled by `try_set_reg` which will use the register page or CPU context.

For testing on WHP, it might be worth batching registers here, but:
1. This change isn't regressing anything from TOT if I don't add register batching
2. Our WHP backend in OpenVMM currently converts a multiple-register set-registers hypercall into a series of single-register hypercalls.